src/nsrt_learning/segmentation.py

```diff
@@ -46,6 +51,47 @@ def _switch_fn(t: int) -> bool:
     return _segment_with_switch_function(trajectory, _switch_fn)
 
 
+def _segment_with_oracle(trajectory: GroundAtomTrajectory) -> List[Segment]:
+    """Segment a trajectory using oracle NSRTs.
+
+    If options are known, just uses _segment_with_option_changes().
+
+    Otherwise, starting at the beginning of the trajectory, keeps track of
+    which oracle ground NSRTs are applicable. When any of them have their
+    effects achieved, that marks the switch point between segments.
+    """
+    traj, all_atoms = trajectory
+    if not traj.actions or traj.actions[0].has_option():
+        return _segment_with_option_changes(trajectory)
+    env = get_or_create_env(CFG.env)
+    gt_nsrts = get_gt_nsrts(env.predicates, env.options)
+    ground_nsrts = {
+        ground_nsrt
+        for nsrt in gt_nsrts
+        for ground_nsrt in utils.all_ground_nsrts(nsrt, list(traj.states[0]))
+    }
+    atoms = all_atoms[0]
+    current_nsrts = list(utils.get_applicable_operators(ground_nsrts, atoms))
+
+    def _switch_fn(t: int) -> bool:
+        nonlocal current_nsrts  # update at each switch point
+        atoms = all_atoms[t + 1]
+        # Check if any of the current NSRT effects hold.
+        for ground_nsrt in current_nsrts:
+            # Check effects.
```

Comment:
is it worth having an `prev_atoms = all_atoms[t]; assert ground_nsrt.preconditions.issubset(prev_atoms)` here?

Reply:
i could imagine that assertion not always holding, and I don't think we need it to?

Reply:
how is that possible? didn't you create the current_nsrts list to exactly be the states where the preconditions hold?

oh, i think i see. in the process of "executing" this operator, you may go through some states where the preconditions don't hold (but the effects haven't been reached yet). because the GroundAtomTrajectory is at the level of the low-level actions. is that what you mean?

Reply:
yeah exactly